build-logic/packaging/src/test/kotlin/gradlebuild/packaging/GradleDistributionInstallTest.kt

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gradlebuild.packaging
+
+import org.gradle.testkit.runner.GradleRunner
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+
+
+class GradleDistributionInstallTest {
+    @TempDir
+    private
+    lateinit var temporaryFolder: File
+
+    private
+    lateinit var target: File
+
+    private
+    lateinit var projectRoot: File
+
+    @BeforeEach
+    fun setup() {
+        target = File(temporaryFolder, "target")
+        target.mkdir()
+
+        projectRoot = File(temporaryFolder, "gradle")
+
+        createMinimalDistribution()
+    }
+
+    @Test
+    fun `installs into empty dir`() {
+        assertSucceeds()
+    }
+
+    @Test
+    fun `installs to non-existing dir`() {
+        target = File(target, "non-existing")
+
+        assertSucceeds()
+    }
+
+    @Test
+    fun `installs into previous distribution`() {
+        assertSucceeds()
+        assertSucceeds()
+    }
+
+    @Test
+    fun `installs into something that looks like previous distribution`() {
+        target.resolve("bin").apply {
+            mkdir()
+            File(this, "gradle").writeText("stub")
+            File(this, "gradle.exe").writeText("stub")
+        }
+
+        target.resolve("lib").apply {
+            mkdir()
+            File(this, "gradle-8.0.2.jar").writeText("stub")
+            File(this, "all-deps-in-the-world-1.2.2.jar").writeText("stub")
+        }
+
+        assertSucceeds()
+    }
+
+    @Test
+    fun `does not install to a file`() {
+        val file = File(target, "some_file.txt").also {
+            it.writeText("some content")
+        }
+        target = file
+
+        assertFails("Install directory $file is not valid: it is actually a file")
+    }
+
+    @Test
+    fun `does not install to non-empty dir without lib`() {
+        target.resolve("bin").apply {
+            mkdir()
+            File(this, "gradle").writeText("stub")
+            File(this, "gradle.exe").writeText("stub")
+        }
+
+        assertFails("Install directory $target does not look like a Gradle installation. Cannot delete it to install.")
```

Comment:
:nerd_face: It would be nice to assert that the `target` directory and its content was not deleted. Same comment for all tests where the installation is aborted.